Upgrade to brick-1.* - EventM API changes

I'm working on an overhaul of `brick`'s `EventM` API and I'd love to get your feedback on how it might impact this project. Please see the details and post comments on https://github.com/jtdaugherty/brick/issues/379. Thanks!

Thanks for the ping. This project needs some adjustments with new brick:

```
[2 of 8] Compiling Summoner.Tui.Field

src/Summoner/Tui/Field.hs:103:46: error:
    • Expecting one more argument to ‘EventM n Bool’
      Expected a type, but ‘EventM n Bool’ has kind ‘* -> *’
    • In the type signature:
        handleEvent :: BrickEvent n e -> Bool -> EventM n Bool
      In an equation for ‘activeCheckboxField’:
          activeCheckboxField stLens isActive name label initialState
            = FormFieldState
                {formFieldState = initVal, formFields = [checkboxFormField],
                 formFieldLens = stLens, formFieldRenderHelper = id,
                 formFieldConcat = vBox}
            where
                initVal, isEnabled :: Bool
                initVal = initialState ^. stLens
                isEnabled = isActive initialState name
                handleEvent :: BrickEvent n e -> Bool -> EventM n Bool
                ....
    |
103 |     handleEvent :: BrickEvent n e -> Bool -> EventM n Bool
    |                                              ^^^^^^^^^^^^^
[3 of 8] Compiling Summoner.Tui.Kit
[4 of 8] Compiling Summoner.Tui.Widget

src/Summoner/Tui/Widget.hs:17:21: error:
    Module ‘Brick.Types’ does not export ‘Padding’
   |
17 | import Brick.Types (Padding (Pad), Widget)
   |                     ^^^^^^^^^^^^^
```

--- a/.github/CODEOWNERS
+++ b/.github/CODEOWNERS
@@ -1,1 +1,1 @@
-*  @chshersh @vrom911
+*  @vrom911

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -86,19 +86,22 @@ jobs:
         echo "packages: ." > cabal.project
         cabal build all
 
-    - name: Build 'cabal-full'
+    - if: matrix.ghc == '9.0.2'
+      name: Build 'cabal-full'
       run: |
         cd "summoner-cli/examples/cabal-full/"
         echo "packages: ." > cabal.project
         cabal build all
 
-    - name: Build 'stack-full'
+    - if: matrix.ghc == '9.0.2'
+      name: Build 'stack-full'
       run: |
         cd "summoner-cli/examples/stack-full/"
         echo "packages: ." > cabal.project
         cabal build all
 
-    - name: Build 'full-batteries'
+    - if: matrix.ghc == '9.0.2'
+      name: Build 'full-batteries'
       run: |
         cd "summoner-cli/examples/full-batteries/"
         echo "packages: ." > cabal.project

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -23,6 +23,7 @@ jobs:
           - "8.8.4"
           - "8.10.7"
           - "9.0.2"
+          - "9.2.4"
         exclude:
           - os: macOS-latest
             ghc: 8.10.7
@@ -43,9 +44,9 @@ jobs:
             ghc: 8.4.4
 
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
 
-    - uses: haskell/actions/setup@v1.2
+    - uses: haskell/actions/setup@v2
       id: setup-haskell-cabal
       name: Setup Haskell
       with:
@@ -60,7 +61,7 @@ jobs:
       run: |
         cabal freeze
 
-    - uses: actions/cache@v2
+    - uses: actions/cache@v3
       name: Cache ~/.cabal/store
       with:
         path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}
@@ -112,15 +113,15 @@ jobs:
         ghc: ["9.0.2"]
 
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
 
-    - uses: haskell/actions/setup@v1
+    - uses: haskell/actions/setup@v2
       name: Setup Haskell Stack
       with:
         ghc-version: ${{ matrix.ghc }}
         stack-version: ${{ matrix.stack }}
 
-    - uses: actions/cache@v1
+    - uses: actions/cache@v3
       name: Cache ~/.stack
       with:
         path: ~/.stack

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,10 @@ The changelog is available [on GitHub][2].
 
 ## Unreleased
 
+* [#564](https://github.com/kowainik/summoner/issues/564):
+  Upgrade `brick` to version `>=1.0`.
+* [#565](https://github.com/kowainik/summoner/issues/565):
+  Build Summoner with GHC-9.2.
 * [#539](https://github.com/kowainik/summoner/issues/539):
   Use GHC-9.0 as default version for generated projects.
 * [#540](https://github.com/kowainik/summoner/issues/540):

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,4 +1,4 @@
-resolver: nightly-2022-02-18
+resolver: lts-19.25
 
 packages:
   - summoner-cli/

--- a/stack.yaml
+++ b/stack.yaml
@@ -5,9 +5,13 @@ packages:
   - summoner-tui/
 
 extra-deps:
-  - colourista-0.1.0.1
+  - brick-1.3
+  - bimap-0.5.0 # for brick
+  - text-zipper-0.12 # for brick
+  - colourista-0.1.0.2
   - megaparsec-9.0.1
   - microaeson-0.1.0.0
-  - tomland-1.3.3.0
+  - tomland-1.3.3.2
   - validation-selective-0.1.0.1
+  - vty-5.37
   - selective-0.4.2

--- a/summoner-cli/summoner.cabal
+++ b/summoner-cli/summoner.cabal
@@ -21,6 +21,7 @@ tested-with:         GHC == 8.4.4
                      GHC == 8.8.4
                      GHC == 8.10.7
                      GHC == 9.0.2
+                     GHC == 9.2.4
 extra-source-files:
     examples/summoner-default.toml
 
@@ -73,8 +74,8 @@ source-repository head
   location: git@github.com:kowainik/summoner.git
 
 common common-options
-  build-depends:       base >= 4.11 && < 4.16
-                     , relude ^>= 1.0.0.0
+  build-depends:       base >= 4.11 && < 4.17
+                     , relude >= 1.0.0.0 && < 1.2
 
   mixins:              base hiding (Prelude)
                      , relude (Relude as Prelude)
@@ -188,7 +189,7 @@ test-suite summoner-test
 
   build-depends:       directory
                      , filepath
-                     , hedgehog >= 0.5.3 && < 1.2
+                     , hedgehog >= 0.5.3 && < 1.3
                      , hspec >= 2.4.8
                      , hspec-hedgehog ^>= 0.0.1.1
                      , tomland

--- a/summoner-tui/src/Summoner/Tui.hs
+++ b/summoner-tui/src/Summoner/Tui.hs
@@ -16,13 +16,13 @@ module Summoner.Tui
        ( summonTui
        ) where
 
-import Brick (App (..), AttrMap, BrickEvent (..), Widget, attrMap, continue, customMain, halt,
+import Brick (App (..), AttrMap, BrickEvent (..), Widget, attrMap, attrName, customMain, halt,
               simpleApp, str, txt, vBox, withAttr, (<+>))
 import Brick.Focus (focusRingCursor)
 import Brick.Forms (allFieldsValid, focusedFormInputAttr, formFocus, formState, handleFormEvent,
                     invalidFormInputAttr, renderForm)
 import Brick.Main (ViewportScroll, neverShowCursor, vScrollBy, viewportScroll)
-import Brick.Types (EventM, Next, ViewportType (Vertical))
+import Brick.Types (EventM, ViewportType (Vertical))
 import Brick.Util (fg)
 import Brick.Widgets.Border (borderAttr)
 import Brick.Widgets.Center (center)
@@ -122,49 +122,51 @@ runTuiNew kit = do
 appNew :: [FilePath] -> App (KitForm e) e SummonForm
 appNew dirs = App
     { appDraw = drawNew dirs
-    , appHandleEvent = \s ev -> if formState s ^. shouldSummon == Idk
+    , appHandleEvent = \ev -> do
+        s <- get
+        if formState s ^. shouldSummon == Idk
         then case ev of
-            VtyEvent (V.EvKey V.KEnter []) -> halt $ changeShouldSummon Yes s
-            VtyEvent (V.EvKey V.KEsc [])   -> withForm ev s (changeShouldSummon Nop)
-            _otherKey                      -> continue s
+            VtyEvent (V.EvKey V.KEnter []) -> put (changeShouldSummon Yes s) >> halt
+            VtyEvent (V.EvKey V.KEsc [])   -> withForm ev (changeShouldSummon Nop)
+            _otherKey                      -> pass
         else case ev of
-            VtyEvent V.EvResize {} -> continue s
+            VtyEvent V.EvResize {} -> pass
             VtyEvent (V.EvKey V.KEnter [V.MMeta]) ->
-                withForm ev s (validateForm . projectDescNewLine)
+                withForm ev (validateForm . projectDescNewLine)
             VtyEvent (V.EvKey V.KEnter []) ->
                 if allFieldsValid s
-                then withForm ev s (changeShouldSummon Idk)
-                else continue s
-            VtyEvent (V.EvKey V.KEsc []) -> halt s
+                then withForm ev (changeShouldSummon Idk)
+                else pass
+            VtyEvent (V.EvKey V.KEsc []) -> halt
             VtyEvent (V.EvKey (V.KChar 'd') [V.MCtrl]) ->
-                withForm ev s (validateForm . ctrlD)
+                withForm ev (validateForm . ctrlD)
 
             -- Handle active/inactive checkboxes
             VtyEvent (V.EvKey (V.KChar ' ') []) -> case getCurrentFocus s of
-                Nothing    -> withFormDef ev s
-                Just field -> handleCheckboxActivation ev s field
+                Nothing    -> withFormDef ev
+                Just field -> handleCheckboxActivation ev field
 
             -- Run handler for autofill actions
             VtyEvent (V.EvKey key [])
                 | keyTriggersAutofill key
-                -> withForm ev s (validateForm . handleAutofill)
+                -> withForm ev (validateForm . handleAutofill)
 
-            MouseDown n _ _ _ -> handleCheckboxActivation ev s n
+            MouseDown n _ _ _ -> handleCheckboxActivation ev n
 
             -- Handle skip of deactivated checkboxes
-            VtyEvent (V.EvKey (V.KChar '\t') []) -> loopWhileInactive ev s
-            VtyEvent (V.EvKey V.KBackTab     []) -> loopWhileInactive ev s
+            VtyEvent (V.EvKey (V.KChar '\t') []) -> loopWhileInactive ev
+            VtyEvent (V.EvKey V.KBackTab     []) -> loopWhileInactive ev
 
             -- Default action
-            _otherEvent -> withFormDef ev s
+            _otherEvent -> withFormDef ev
 
     , appChooseCursor = focusRingCursor formFocus
-    , appStartEvent = pure
+    , appStartEvent = pass
     , appAttrMap = const theMap
     }
   where
-    withForm  ev s f = handleFormEvent ev s >>= continue . f
-    withFormDef ev s = withForm ev s validateForm
+    withForm ev f = handleFormEvent ev >> modify f
+    withFormDef ev = withForm ev validateForm
 
     changeShouldSummon :: Decision -> KitForm e -> KitForm e
     changeShouldSummon newShould f = mkForm $ formState f & shouldSummon .~ newShould
@@ -178,29 +180,28 @@ appNew dirs = App
     -- Activate/Deactivate checkboxes depending on current field change
     handleCheckboxActivation
         :: BrickEvent SummonForm e
-        -> KitForm e
         -> SummonForm
-        -> EventM SummonForm (Next (KitForm e))
-    handleCheckboxActivation ev form = \case
-        CabalField     -> withForm ev form mkNewForm
-        StackField     -> withForm ev form mkNewForm
-        GitHubEnable   -> withForm ev form mkNewForm
-        GitHubDisable  -> withForm ev form mkNewForm
-        GitHubNoUpload -> withForm ev form mkNewForm
-        _nonCheckBox   -> withFormDef ev form
+        -> EventM SummonForm (KitForm e) ()
+    handleCheckboxActivation ev = \case
+        CabalField     -> withForm ev mkNewForm
+        StackField     -> withForm ev mkNewForm
+        GitHubEnable   -> withForm ev mkNewForm
+        GitHubDisable  -> withForm ev mkNewForm
+        GitHubNoUpload -> withForm ev mkNewForm
+        _nonCheckBox   -> withFormDef ev
 
     -- Handles form event until current element is active
     loopWhileInactive
         :: BrickEvent SummonForm e
-        -> KitForm e
-        -> EventM SummonForm (Next (KitForm e))
-    loopWhileInactive ev form = do
-        newForm <- handleFormEvent ev form
+        -> EventM SummonForm (KitForm e) ()
+    loopWhileInactive ev = do
+        handleFormEvent ev
+        newForm <- get
         case getCurrentFocus newForm of
-            Nothing -> continue newForm
+            Nothing -> pass
             Just field -> if not $ isActive (formState newForm) field
-                then loopWhileInactive ev newForm
-                else continue newForm
+                then loopWhileInactive ev
+                else pass
 
     -- Autofill is only triggered on characters and backspace keys.
     keyTriggersAutofill :: V.Key -> Bool
@@ -235,7 +236,7 @@ drawNew dirs kitForm = case kit ^. shouldSummon of
 
     tree :: Widget SummonForm
     tree = hLimitPercent 25 $ vLimit 22 $ borderLabel "Project Structure" $ vBox
-        [ withAttr "tree" $ txt $ renderWidgetTree kit
+        [ withAttr (attrName "tree") $ txt $ renderWidgetTree kit
         -- to fill all the space that widget should take.
         , fill ' '
         ]
@@ -257,12 +258,12 @@ drawNew dirs kitForm = case kit ^. shouldSummon of
             _noInfoAvailable -> emptyWidget
 
         infoTxt :: Text -> Widget SummonForm
-        infoTxt = withAttr "blue-fg" . txtWrap . (<>) " ⓘ  "
+        infoTxt = withAttr (attrName "blue-fg") . txtWrap . (<>) " ⓘ  "
 
         validationBlock :: Widget SummonForm
         validationBlock = vBox $ case formErrorMessages dirs kitForm of
-            []     -> [withAttr "green-fg" $ str " ✔  Project configuration is valid"]
-            fields -> map (\msg -> withAttr "red-fg" $ strWrap $ " ☓  " ++ msg) fields
+            []     -> [withAttr (attrName "green-fg") $ str " ✔  Project configuration is valid"]
+            fields -> map (\msg -> withAttr (attrName "red-fg") $ strWrap $ " ☓  " ++ msg) fields
 
         configBlock :: Widget SummonForm
         configBlock = case kit ^. configFile of
@@ -314,14 +315,14 @@ runTuiShowLicense (toText -> name) = case parseLicenseName name of
     licenseApp :: LicenseName -> License -> App () e ()
     licenseApp licenseName lc = App
         { appDraw         = drawScrollableLicense licenseName lc
-        , appStartEvent   = pure
+        , appStartEvent   = pass
         , appAttrMap      = const theMap
         , appChooseCursor = neverShowCursor
-        , appHandleEvent  = \() event -> case event of
-            VtyEvent (V.EvKey V.KDown []) -> vScrollBy licenseScroll   1  >> continue ()
-            VtyEvent (V.EvKey V.KUp [])   -> vScrollBy licenseScroll (-1) >> continue ()
-            VtyEvent (V.EvKey V.KEsc [])  -> halt ()
-            _otherEvent                   -> continue ()
+        , appHandleEvent  = \case
+            VtyEvent (V.EvKey V.KDown []) -> vScrollBy licenseScroll   1  >> pass
+            VtyEvent (V.EvKey V.KUp [])   -> vScrollBy licenseScroll (-1) >> pass
+            VtyEvent (V.EvKey V.KEsc [])  -> halt
+            _otherEvent                   -> pass
         }
 
     licenseScroll :: ViewportScroll ()
@@ -375,10 +376,10 @@ theMap = attrMap V.defAttr
     , (listSelectedAttr,        V.black `Brick.on` V.cyan)
     , (listSelectedFocusedAttr, V.black `Brick.on` V.white)
     , (disabledAttr,            fg V.brightBlack)
-    , ("blue-fg",               fg V.blue)
-    , ("green-fg",              fg V.green)
-    , ("yellow-fg",             fg V.yellow)
-    , ("red-fg",                fg V.brightRed)
+    , (attrName "blue-fg",      fg V.blue)
+    , (attrName "green-fg",     fg V.green)
+    , (attrName "yellow-fg",    fg V.yellow)
+    , (attrName "red-fg",       fg V.brightRed)
     , (borderAttr,              fg V.cyan)
-    , ("tree",                  fg V.cyan)
+    , (attrName "tree",         fg V.cyan)
     ]

--- a/summoner-tui/src/Summoner/Tui/Field.hs
+++ b/summoner-tui/src/Summoner/Tui/Field.hs
@@ -20,8 +20,8 @@ module Summoner.Tui.Field
        , disabledAttr
        ) where
 
-import Brick (BrickEvent (..), EventM, Location (..), Widget, clickable, showCursor, str, vBox,
-              withAttr, withDefAttr, (<+>))
+import Brick (BrickEvent (..), EventM, Location (..), Widget, attrName, clickable, showCursor, str,
+              vBox, withAttr, withDefAttr, (<+>))
 import Brick.AttrMap (AttrName)
 import Brick.Forms (FormField (..), FormFieldState (..), checkboxCustomField, focusedFormInputAttr,
                     radioCustomField)
@@ -105,11 +105,11 @@ activeCheckboxField stLens isActive name label initialState = FormFieldState
     initVal   = initialState ^. stLens
     isEnabled = isActive initialState name
 
-    handleEvent :: BrickEvent n e -> Bool -> EventM n Bool
+    handleEvent :: BrickEvent n e -> EventM n Bool ()
     handleEvent (MouseDown n _ _ _)
-        | isEnabled && n == name = pure . not
-    handleEvent (VtyEvent (V.EvKey (V.KChar ' ') [])) = pure . not
-    handleEvent _ = pure
+        | isEnabled && n == name = get >>= put . not
+    handleEvent (VtyEvent (V.EvKey (V.KChar ' ') [])) = get >>= put . not
+    handleEvent _ = pass
 
     checkboxFormField :: FormField Bool Bool e n
     checkboxFormField = FormField
@@ -132,4 +132,4 @@ renderCheckbox isEnabled label n foc val =
 
 -- | Attribute for disabled checkboxes.
 disabledAttr :: AttrName
-disabledAttr = "disabled"
+disabledAttr = attrName "disabled"

--- a/summoner-tui/src/Summoner/Tui/Widget.hs
+++ b/summoner-tui/src/Summoner/Tui/Widget.hs
@@ -17,11 +17,12 @@ module Summoner.Tui.Widget
        , listInBorder
        ) where
 
-import Brick.Types (Padding (Pad), Widget)
+import Brick (attrName)
+import Brick.Types (Widget)
 import Brick.Widgets.Border (borderWithLabel)
 import Brick.Widgets.Center (center)
-import Brick.Widgets.Core (hBox, hLimit, padLeftRight, padRight, str, txtWrap, vBox, vLimit,
-                           withAttr, (<+>))
+import Brick.Widgets.Core (Padding (Pad), hBox, hLimit, padLeftRight, padRight, str, txtWrap, vBox,
+                           vLimit, withAttr, (<+>))
 
 
 -- | Adds label to the Form's field.
@@ -91,4 +92,4 @@ listInBorder name limitH extraLimitV list = center
     $ borderLabel name
     $ center
     $ vBox
-    $ map (withAttr "blue-fg" . txtWrap . ("➤ " <>)) list
+    $ map (withAttr (attrName "blue-fg") . txtWrap . ("➤ " <>)) list

--- a/summoner-tui/summoner-tui.cabal
+++ b/summoner-tui/summoner-tui.cabal
@@ -21,14 +21,15 @@ tested-with:         GHC == 8.4.4
                      GHC == 8.8.4
                      GHC == 8.10.7
                      GHC == 9.0.2
+                     GHC == 9.2.4
 
 source-repository head
   type:     git
   location: git@github.com:kowainik/summoner.git
 
 common common-options
-  build-depends:       base >= 4.11 && < 4.16
-                     , relude ^>= 1.0.0.0
+  build-depends:       base >= 4.11 && < 4.17
+                     , relude >= 1.0.0.0 && < 1.2
 
   mixins:              base hiding (Prelude)
                      , relude (Relude as Prelude)
@@ -80,14 +81,14 @@ library
   autogen-modules:     Paths_summoner_tui
   other-modules:       Paths_summoner_tui
 
-  build-depends:       brick >= 0.56 && < 0.68
+  build-depends:       brick >= 1.0 && < 1.4
                      , colourista ^>= 0.1
                      , directory ^>= 1.3.0.2
                      , microlens ^>= 0.4
                      , microlens-th ^>= 0.4
                      , summoner ^>= 2.0.0.0
                      , validation-selective >= 0.0.0.0 && < 0.2
-                     , vty >= 5.25 && < 5.34
+                     , vty >= 5.36 && < 5.38
 
 executable summon-tui
   import:              common-options

